### PR TITLE
Filter out excise measures on xi and refactor measure presenter

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1120

### What?

Also see: https://github.com/trade-tariff/trade-tariff-backend/pull/300

I have added/removed/altered:

- [x] Rename MeasurePresenter -> MeasureCollection and give it a collection/scope interface
- [x]  Update consumers of the MeasureCollection implementation
- [x] Adds filtering to the MeasureCollection class to exclude excise measures when running the xi service
- [x] Adds MeasureCollection spec

### Why?

I am doing this because:

- We're being bitten by the fact that we have excise measures from CHIEF that are no longer maintained and could do with being end dated at some stage. these measures should be filtered out in XI

### CODE REVIEW

File: app/controllers/api/v1/commodities_controller.rb [hunk 1]
```diff
@@ -32,7 +32,7 @@ def show
             :full_temporary_stop_regulations,
             :measure_partial_temporary_stops
           ).all, @commodity
-        ).validate!
+        ).deduplicate.filter
```
[the remark this answers is not in the document]
Comment (alexdesi) [reply app/controllers/api/v1/commodities_controller.rb:35]: what about renaming 
`deduplicate` to  `uniq`
shorter and simpler :)

File: spec/models/measure_collection_spec.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  subject(:collection) { described_class.new(measures, declarable) }
+
+  let(:declarable) { build(:commodity) }
+
+  describe '#filter' do
+    before do
+      allow(TradeTariffBackend).to receive(:service).and_return(service)
+    end
+
+    let(:service) { 'uk' }
```
Comment (alexdesi) [inline spec/models/measure_collection_spec.rb:11]: do we need this?

File: app/models/measure_collection.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        measure.measure_type_id == THIRD_COUNTRY_DUTY_ID && measure.additional_code.blank? && measure.goods_nomenclature_sid != @declarable.goods_nomenclature_sid
+      }.compact!
+    end
+  end
+
+  def delete_duplicate_vat_measures(type:)
+    return unless select { |measure| measure.measure_type_id == type }.any?
+
+    delete_if { |measure| measure.measure_type_id == type && measure.goods_nomenclature_sid != @declarable.goods_nomenclature_sid }
+    latest = select { |measure| measure.measure_type_id == type }.sort_by(&:effective_start_date).pop
```
Comment (alexdesi) [inline app/models/measure_collection.rb:49]: pop implies we are removing the last element, but is't not clear from where,
does not select create a copy of the collection?
should you use last? 🤷

File: app/models/measure_collection.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  def third_country_duty_dedup
+    # TODO: Fix measures with additional codes that are duplicate
+    if select { |measure| measure.measure_type_id == THIRD_COUNTRY_DUTY_ID }.size > 1
+      delete_if { |measure|
+        measure.measure_type_id == THIRD_COUNTRY_DUTY_ID && measure.additional_code.blank? && measure.goods_nomenclature_sid != @declarable.goods_nomenclature_sid
+      }.compact!
+    end
+  end
+
+  def delete_duplicate_vat_measures(type:)
```
Comment (alexdesi) [inline app/models/measure_collection.rb:45]: delete_duplicate is a synonym for `unique`, so we could rename it to 
`vat_measure_unique!`
up to you

File: app/models/measure_collection.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    return unless select { |measure| measure.measure_type_id == type }.any?
+
+    delete_if { |measure| measure.measure_type_id == type && measure.goods_nomenclature_sid != @declarable.goods_nomenclature_sid }
+    latest = select { |measure| measure.measure_type_id == type }.sort_by(&:effective_start_date).pop
+    delete_if { |measure| measure.measure_type_id == type && matching_item_id(measure) }
+    prepend(latest) unless include?(latest)
+    compact!
+  end
```
Comment (alexdesi) [inline app/models/measure_collection.rb:53]: this is a bit complex, it  is filtering by measure_type_id 4 times, I am wondering if we can simplify it

`{ |measure| measure.measure_type_id == type }`

File: spec/models/measure_collection_spec.rb (hunk 2 above)
Comment (willfish) [reply spec/models/measure_collection_spec.rb:11]: See line 8 and the one spec that isn't interested in the service

File: app/models/measure_collection.rb (hunk 3 above)
Comment (willfish) [reply app/models/measure_collection.rb:49]: This is copied and pasted for the most part. I'm planning on having another pop at this file later

File: app/models/measure_collection.rb (hunk 5 above)
Comment (willfish) [reply app/models/measure_collection.rb:53]: This is copied and pasted for the most part. The main bits I've changed is the interface and the addition of the filter method.

**Review by alexdesi — APPROVED** [review]

**Comment by alexdesi** [comment]
have a look to the linters issue, they seems easy to fix

**Pull request merged**